File: Csharp/Benchmark/Benchmark/Mandelbrot.cs
using System.Numerics;
using SkiaSharp;

namespace Benchmark.Tasks
{

    public enum MandelbrotMethod
    {
        Sequential,
        Parallel
    }

    /// <summary>
    /// Generate Mandelbrot sets
    /// </summary>
    public class Mandelbrot
    {

        #region properties

        int Width { init; get; }
        int Height { init; get; }
        double XCenter { init; get; }
        double YCenter { init; get; }
        double Scale { init; get; }

        #endregion

        #region private members

        private static readonly double R_MAX = 2.0;
        private static readonly byte N_MAX = 255;

        private SKImage img;
        //private SKBitmap bm;

        #endregion

        #region static members

        /// <summary>
        /// Calculate pixel value for given complex number
        /// </summary>
        /// <param name="z"></param>
        /// <returns>Pixel value</returns>
        public static byte PixelValue(Complex z)
        {
            var n = N_MAX;
            var z0 = z;

            while (z.Magnitude < R_MAX && n > 0)
            {
                z = z * z + z0;
                --n;
            }

            return n;
        }

        /// <summary>
        /// Calculate pixel value for given coordinates
        /// </summary>
        /// <param name="x">Horizontal coordinate</param>
        /// <param name="y">Vertical coordinate</param>
        /// <returns>Pixel value</returns>
        public static byte PixelValue(double x, double y)
        {
            return PixelValue(new Complex(x, y));
        }

        /// <summary>
        /// Convert pixel value to RGB values
        /// </summary>
        /// <param name="value">Pixel value</param>
        /// <returns>Byte array with RGB values</returns>
        public static byte[] PixelToRgb(byte value)
        {
            var r = (byte)(15 * (value % 15));
            var g = (byte)(32 * (value % 7));
            var b = (byte)(8 * (value % 31));
            var rgb = new byte[] { r, g, b };

            return rgb;
        }

        #endregion

        #region constructor

        /// <summary>
        /// Construct a Mandelbrot set
        /// </summary>
        /// <param name="width">Image width of Mandelbrot set</param>
        /// <param name="height">Image height of Mandelbrot set</param>
        /// <param name="xCenter">Horizontal center coordinate of Mandelbrot set</param>
        /// <param name="yCenter">Vertical center coordinate of Mandelbrot set</param>
        /// <param name="scale">Pixel size scaling</param>
        public Mandelbrot(int width, int height, double xCenter, double yCenter, double scale, MandelbrotMethod method)
        {
            Width = width;
            Height = height;
            XCenter = xCenter;
            YCenter = yCenter;
            Scale = scale;

            var info = new SKImageInfo(Width, Height, SKColorType.Rgb888x);

            var buf = method switch
            {
                MandelbrotMethod.Sequential => CalcPixelSeq(),
                MandelbrotMethod.Parallel => CalcPixelPar()
            };

            img = SKImage.FromPixelCopy(info, buf);
        }

        #endregion

        #region methods

        /// <summary>
        /// Calculate Mandelbrot set sequentially
        /// </summary>
        /// <param name="data">Byte array of RGB pixel data</param>
        public Span<byte> CalcPixelSeq()
        {
            var xOffset = XCenter - 0.5 * Scale * Width;
            var yOffset = YCenter + 0.5 * Scale * Width;

            byte[] buf = new byte[4 * Width * Height];

            for (int y = 0; y < Height; y++)
            {
                for (int x = 0; x < Width; x++)
                {
                    var xCoord = x * Scale + xOffset;
                    var yCoord = y * Scale - yOffset;

                    var pixel = PixelValue(xCoord, yCoord);
                    var rgb = PixelToRgb(pixel);

                    // set RGB values, leave alpha out
                    var baseIdx = 4 * x + 4 * Width * y;
                    buf[baseIdx] = rgb[0];
                    buf[baseIdx + 1] = rgb[1];
                    buf[baseIdx + 2] = rgb[2];
                }
            }
            return buf;
        }

        public Span<byte> CalcPixelPar()
        {
            var xOffset = XCenter - 0.5 * Scale * Width;
            var yOffset = YCenter + 0.5 * Scale * Width;

            byte[] buf = new byte[4 * Width * Height];

            Parallel.For(0, Height, y =>
            {
                for (int x = 0; x < Width; x++)
                {
                    var xCoord = x * Scale + xOffset;
                    var yCoord = y * Scale - yOffset;

                    var pixel = PixelValue(xCoord, yCoord);
                    var rgb = PixelToRgb(pixel);

                    // set RGB values, leave alpha out
                    var baseIdx = 4 * x + 4 * Width * y;
                    buf[baseIdx] = rgb[0];
                    buf[baseIdx + 1] = rgb[1];
                    buf[baseIdx + 2] = rgb[2];
                }
            });
            return buf;
        }

        /// <summary>
        /// Save Mandelbrot image to file
        /// </summary>
        /// <param name="filename">File path to be written</param>
        public void Save(string filename)
        {
            using (Stream s = File.Create(filename))
            {
                var d = img.Encode(SKEncodedImageFormat.Png, 100);
                d.SaveTo(s);
            }
        }

        #endregion
    }
}

File: Python/benchmark/src/benchmark/mandelbrot.py
# mandelbrot.py
# Generate Mandelbrot set images in Python (V3)

from PIL import Image


class Mandelbrot:
    """Class to generate Mandelbrot set images."""

    def __init__(
        self,
        width: int,
        height: int,
        center: tuple[float, float],
        pixel_size: float,
        max_iter: int,
    ):
        self.width = width
        self.height = height
        self.pixel_size = pixel_size
        self.x_offset = center[0] - 0.5 * pixel_size * self.width
        self.y_offset = center[1] + 0.5 * pixel_size * self.height
        self.max_iter = max_iter
        self.image = Image.new("RGB", (self.width, self.height))

    def _pixel_value(
        self,
        z0: complex,
    ) -> int:
        """Calculate the number of iterations for a point in the Mandelbrot set."""
        n = 0
        z = complex(0, 0)
        while abs(z) <= 2 and n < self.max_iter:
            z = z * z + z0
            n += 1
        return n

    def _get_color(self, n: int) -> tuple[int, int, int]:
        """Return a color based on the number of iterations."""
        if n >= self.max_iter:
            return (0, 0, 0)  # Black for points in the set
        # return (n % 8 * 32, n % 16 * 16, n % 32 * 8)  # Color based on iteration count
        return (
            5 * (n % 15),
            16 * (n % 7),
            8 * (n % 31),
        )  # Color based on iteration count

    def generate(self):
        """Generate the Mandelbrot set image."""
        for x in range(self.width):
            for y in range(self.height):
                x_scaled = x
                y_scaled = y
                z = complex(x_scaled, y_scaled)
                pix_val = self._pixel_value(z)
                color = self._get_color(pix_val)

                zx = (x / self.width) * 3.5 - 2.5
                zy = (y / self.height) * 2.0 - 1.0
                z0 = complex(zx, zy)
                n = self._pixel_value(z0)
                color = self._get_color(n)
                self.image.putpixel((x, y), color)  # slow, alternative: 'paste' method

    def save(self, filename: str):
        """Save the generated image to a file."""
        self.image.save(filename, "PNG")
